src/Validation/Requires.cs

```diff
@@ -189,16 +190,20 @@ public static void NotNullOrEmpty([ValidatedNotNull, NotNull] System.Collections
                 throw new ArgumentNullException(parameterName);
             }
 
-            bool hasElements = false;
-            foreach (object value in values)
+            IEnumerator enumerator = values.GetEnumerator();
+            try
             {
-                hasElements = true;
-                break;
+                if (!values.GetEnumerator().MoveNext())
+                {
+                    throw new ArgumentException(Format(Strings.Argument_EmptyArray, parameterName), parameterName);
+                }
             }
-
-            if (!hasElements)
+            finally
             {
-                throw new ArgumentException(Format(Strings.Argument_EmptyArray, parameterName), parameterName);
+                if (enumerator is IDisposable disposable)
+                {
+                    disposable.Dispose();
+                }
             }
```

Comment:
tip: a more succinct syntax for `try/finally` based disposal in this case can be achieved like this:

```cs
IEnumerator enumerator = values.GetEnumerator();
using (enumerator as IDisposable)
{
    if (!values.GetEnumerator().MoveNext())
    {
        throw new ArgumentException(Format(Strings.Argument_EmptyArray, parameterName), parameterName);
    }
}
```

Reply:
That gets the enumerator twice though. Wouldn't we want to avoid that?

Reply:
Whoops. Well, that bug was *already* in your code. :) I just copied it in. Yes, let's avoid calling `GetEnumerator()` twice by actually *using* your `enumerator` local.

Reply:
Shoot! You're right. It was. ☹️ I missed that when I changed it to the try...finally.